src/ctapipe/reco/impact.py

```diff
@@ -238,76 +348,46 @@ def get_shower_max(self, source_x, source_y, core_x, core_y, zen):
         float: Depth of maximum of air shower
 
         """
-
         # Calculate displacement of image centroid from source position (in
         # rad)
         disp = np.sqrt((self.peak_x - source_x) ** 2 + (self.peak_y - source_y) ** 2)
+
         # Calculate impact parameter of the shower
         impact = np.sqrt(
             (self.tel_pos_x - core_x) ** 2 + (self.tel_pos_y - core_y) ** 2
         )
         # Distance above telescope is ratio of these two (small angle)
 
-        height = impact / disp
+        height_tilted = (
+            impact / disp
+        )  # This is in tilted frame along the telescope axis
         weight = np.power(self.peak_amp, 0.0)  # weight average by sqrt amplitude
         # sqrt may not be the best option...
 
-        # Take weighted mean of estimates
-        mean_height = np.sum(height * weight) / np.sum(weight)
-        # This value is height above telescope in the tilted system,
-        # we should convert to height above ground
-        mean_height *= np.cos(zen)
+        # Take weighted mean of estimates, converted to height above ground
+        mean_height_above_ground = np.sum(
+            height_tilted * np.cos(zen) * weight
+        ) / np.sum(weight)
 
         # Add on the height of the detector above sea level
-        mean_height += 2150
+        mean_height_asl = (
+            mean_height_above_ground
+            + self.subarray.reference_location.height.to_value(u.m)
+        )
 
-        if mean_height > 100000 or np.isnan(mean_height):
-            mean_height = 100000
+        if mean_height_asl > 100000 or np.isnan(mean_height_asl):
+            mean_height_asl = 100000
```

Comment:
why replace nan or nonsensical values with a valid looking value? That might be misleading. Also a magic number.